### PR TITLE
Term frequency - exponential decay

### PR DESCRIPTION
**Pull request by BeritJanssen**
This is my throw at trying to implement the speedup as suggested in #959 .

One question: in `term_frequency.get_match_count`, I decided to replace `es.scroll` by `es.search`, as - for as far as I understood the code - there were  never more than 5'000 documents fetched and I'm not sure scroll speeds up retrieval in that case, especially since we only need ids and no source data. OR was this solved with `scroll` in order to allow complete data analysis here?

I solved the estimation with sklearn's `auc` function (which estimates the Area under the Curve using the trapezoidal rule), not sure how crude the estimate from this is. Alternatively, one could fit a curve along the points and then sum all y coordinates along this estimated curve, as suggested - with the potential problem that the fitting algorithm may not always converge.

### CODE REVIEW

**Comment by lukavdplas** [comment]
Almost forgot:

- The [manual page](https://github.com/UUDigitalHumanitieslab/I-analyzer/blob/develop/frontend/src/assets/manual/en-GB/termfrequency.md) should be updated accordingly, as well as the warning text about sampling in the barchart and history components.
- Since this introduces a more sophisticated method of sampling, the actual sample size can probably go down. But you can also leave that for a later change if you want to compare results first.

**Comment by lukavdplas** [comment]
> > One question: in term_frequency.get_match_count, I decided to replace es.scroll by es.search, as - for as far as I understood the code - there were never more than 5'000 documents fetched and I'm not sure scroll speeds up retrieval in that case, especially since we only need ids and no source data. OR was this solved with scroll in order to allow complete data analysis here?
> 
> I think it was changed to `scroll` at a time when the term frequency graph analysed all the data. Indeed, as long as the document limit is less than 10.000, you can also use `search`.

Correction on this: the full data download uses the same function, and _that_ is why es.scroll is used.

There should be a unit tests on the large mock corpus that will fail based on this change (provided you have an elasticsearch server) - if not, could you add one?

File: backend/visualization/term_frequency.py [hunk 1]
```diff
@@ -61,23 +68,39 @@ def extract_data_for_term_frequency(corpus, es_query):
...
     query_text = query.get_query_text(es_query)
 
-    matches = sum(
-        count_matches_in_document(hit['_id'], index, fieldnames, query_text, es_client)
-        for hit in found_hits
-    )
+    matches = [
+        count_matches_in_document(found_hits[sample_index]['_id'], index, fieldnames, query_text, es_client)
+        for sample_index in sample_points
+    ]
```
[the remark this answers is not in the document]
Comment (BeritJanssen) [reply backend/visualization/term_frequency.py]: I tried refactoring as you suggested, but found that the doubling of the retrieve_matches (which needs a lot of arguments, namely also the client and found_hits) made the whole thing more verbose and less readable. I compromised by making the if-else clause to determine the sampling points a one-liner.

**Review by lukavdplas — APPROVED** [review]

**Comment by BeritJanssen** [comment]
After a fair amount of experimentation & bugfixes, I realized that the estimates are way off. This is because there is not actual an exponential decay in the number of matches per document - rather, the number of matches tend to fluctuate between [1, 3] in most cases for a typical corpus, nor will documents appear ordered from many too few matches. This is because relevance does a TF-IDF weighting, so longer documents will be penalized and may appear much lower in the ranked list. In consequence, the earlier approach - taking a certain number of documents as sample and assume all other documents to contain the search term at least once - probably comes much closer to the true number.
Query "people" in `parliament-canada`, sampling approach:
Screenshot 2023-08-03 at 14 03 31
Query "people" in `parliament-canada`, estimate approach:
Screenshot 2023-08-03 at 14 03 10

Will open a new PR which just adds some improvements to the current (sample based) implementation:
- we should make sure there is no `sort` part in the queryModel (i.e., we want to sort by relevance, not by date, which is the default for the parliament corpora)
- make sure we don't request the whole shebang of data in the search query, but only document ids
- decrease sample size

**Pull request closed without merging**